rj_gameplay/reg_pass_costing/cost_heuristic.py

```diff
@@ -0,0 +1,157 @@
+import stp.rc as rc
+import rj_gameplay.gameplay_node as gameplay_node
+import rclpy
+import numpy as np
+
+MAX_PASS_DIST = 5
+MAX_GOAL_DIST = 7.5
+MAX_AVG_OPP_DIST = 6
+MAX_NEAREST_OPP_DIST = 2.5
+MAX_OPP_BETWEEN_ROBOT_GOAL = 5
+MAX_OPP_BETWEEN_ROBOT_BALL = 1
+MIN_GOAL_DIST_BONUS = 1
+MAX_GOAL_DIST_BONUS = 1.75
+
+GOAL_DIST_COST = 150
+AVG_OPP_DIST_COST = 75
+NEAREST_OPP_DIST_COST = 300
+OPP_BETWEEN_ROBOT_GOAL_COST = 325
+OPP_BETWEEN_ROBOT_BALL_COST = 650
+
+MAX_COST = 1000
+
+def cost_heuristic(world_state: rc.WorldState, robot: rc.Robot) -> float:
+
+    field = world_state.field
+
+    our_robots = world_state.our_robots
+    their_robots = world_state.their_robots
+
+    robot_with_ball = our_robots[0]
+    for bot in our_robots:
+        if bot.has_ball_sense:
+            robot_with_ball = bot
+            if robot_with_ball == robot:
+                return -1
+
+    # return 1 if robot is too far away from robot_with_ball
+    if (distance_from_ball(robot, robot_with_ball) > MAX_PASS_DIST):
+        return 1 
+
+    running_cost = 0
+    running_cost += distance_from_goal(robot, robot_with_ball, field)
+    running_cost += average_opponent_distance(robot, their_robots)
+    running_cost += nearest_opponent_distance(robot, their_robots)
+    running_cost += opponents_between_robot_goal(robot, their_robots, field)
+    running_cost += opponents_between_robot_ball(robot, their_robots, robot_with_ball)
+
+    if running_cost >= MAX_COST:
+        return 1
+    return round(running_cost / MAX_COST, 3)
+
+# returns distance between the robot and robot_with_ball
+def distance_from_ball(robot: rc.Robot, robot_with_ball: rc.Robot) -> float:
+    robot_pos = robot.pose[0:2]
+    robot_with_ball_pos = robot_with_ball.pose[0:2]
+    return np.linalg.norm(robot_with_ball_pos - robot_pos)
+
+# returns whether or not the opponent robot can intercept pass/path to goal
+def intercept_calc(A: np.array, B: np.array, E: np.array, target_dist: float) -> bool:
+    AB_BE = np.dot(B - A, E - B)
+    AB_AE = np.dot(B - A, E - A)
+
+    if AB_BE < 0 and AB_AE > 0:
+        intercept_dist = abs(np.cross(B - A, E - A)) / np.linalg.norm(B - A)
+        if intercept_dist < target_dist / 6:
+            return True
+    return False
+
+# returns the cost for the robot's distance from goal
+def distance_from_goal(robot: rc.Robot, robot_with_ball: rc.Robot, field: rc.Field) -> float:
```

Comment:
You may want to rename this function, since it is not only distance from goal because of the normalization with the Max and the multiplication of the bonus. The function itself seems good, the name is just a bit misleading.